Implement public API InferenceMode and its error handling

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54403 Relax some limitations of InferenceMode.
* **#53343 Implement public API InferenceMode and its error handling**

RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

Commit:
Implement public API AutoInferenceMode and its error handling

[ghstack-poisoned]

Commit:
Update on "[WIP]Implement public API AutoInferenceMode and its error handling"




[ghstack-poisoned]

Commit:
Update on "[WIP]Implement public API AutoInferenceMode and its error handling"




[ghstack-poisoned]

Commit:
Update on "[WIP]Implement public API AutoInferenceMode and its error handling"




[ghstack-poisoned]

Commit:
Update on "[WIP]Implement public API AutoInferenceMode and its error handling"




[ghstack-poisoned]

Commit:
Update on "[WIP]Implement public API AutoInferenceMode and its error handling"




[ghstack-poisoned]

Commit:
Update on "[WIP]Implement public API AutoInferenceMode and its error handling"




[ghstack-poisoned]

Commit:
Update on "[WIP]Implement public API AutoInferenceMode and its error handling"




[ghstack-poisoned]

Commit:
Update on "[WIP]Implement public API AutoInferenceMode and its error handling"




[ghstack-poisoned]

Commit:
Update on "[WIP]Implement public API AutoInferenceMode and its error handling"




[ghstack-poisoned]

Commit:
Update on "[WIP]Implement public API AutoInferenceMode and its error handling"




[ghstack-poisoned]

ezyang commented on c10/core/inference_mode.cpp:
```diff
@@ -0,0 +1,36 @@
+/// is on.
+#if !defined(C10_MOBILE) || defined(FEATURE_TORCH_MOBILE)
+
+thread_local bool InferenceMode_enabled = false;
```
You sure about this being its own, separate thread local? Aren't you doubling the number of necessary TLS accesses now?

ezyang commented on c10/core/inference_mode.h:
```diff
@@ -0,0 +1,66 @@
+#pragma once
```
bikeshed: can we name this file after the name of the main class it exports, in the same way the other files in this directory are named? :P

ezyang commented on c10/core/inference_mode.h:
```diff
@@ -0,0 +1,66 @@
+
+// A RAII, thread local (!) guard that enables or disables inference mode upon
+// construction, and sets it back to the original value upon destruction.
+struct TORCH_API AutoInferenceMode {
```
Bikeshed: I'd actually vote for naming this simply InferenceMode (and renaming the struct on line 8 to something else, or maybe even just stuffing the static methods on the same class). I don't think you need the "Auto" because "Mode" should already make it clear that an RAII-ey thing is happening here.

At the *very least*, there shouldn't be an attractive nuisance InferenceMode struct in public interface, because that makes it super easy for someone to write:

```
c10::InferenceMode mode;
```

and it would work! It would compile! Your tests would even pass! But you wouldn't actually have gotten the mode.

ezyang commented on aten/src/ATen/templates/TensorBody.h:
```diff
@@ -592,7 +593,8 @@ class TORCH_API Tensor {
   Tensor& set_requires_grad(bool requires_grad) {
-    impl_->set_requires_grad(requires_grad);
+    // requires_grad cannot be set to true in InferenceMode.
+    impl_->set_requires_grad(!c10::InferenceMode::is_enabled() && requires_grad);
```
Is this the right spot to do it? Perhaps this should happen on the TensorImpl?

Commit:
Update on "Implement public API InferenceMode and its error handling"




[ghstack-poisoned]

Commit:
Update on "Implement public API InferenceMode and its error handling"




[ghstack-poisoned]

ezyang commented on c10/core/InferenceMode.h:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  static bool is_enabled();
+  static void set_enabled(bool enabled);
```
This looks like it should be private to me, otherwise someone can do InferenceMode::set_enabled() but this won't actually do anything because the dispatch key TLS won't get updated.

Commit:
Update on "Implement public API InferenceMode and its error handling"


Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

albanD commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -335,6 +342,19 @@ def emit_inplace_or_view_body(fn: NativeFunctionWithDifferentiabilityInfo) -> Li
+    # There're two valid cases calling InplaceOrView:
+    #  - redispatch from VariableType kernel (AutoNonVariableTypeMode is on)
+    #  - In inference mode where Autograd keys are excluded as well.
+    # If InplaceOrView is hit without Autograd keys being disabled, e.g. apply
```
Sentences here are wrong I think? They start with If but lead to nothing

albanD commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -335,6 +342,19 @@ def emit_inplace_or_view_body(fn: NativeFunctionWithDifferentiabilityInfo) -> Li
+    # their base tensor doesn't have complete information about view/version).
+    # This is possible to do, but we currently don't see such use cases so
+    # simply throwing errors here.
+    if len(f.func.returns):
```
Why this check?

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -658,6 +662,18 @@ def emit_any_requires_grad() -> List[str]:
 
+    def emit_assert_no_inference_tensor() -> List[str]:
+        tensor_args = []
+        for arg in f.func.arguments.non_out:
```
Why do you exclude out args here?

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -658,6 +662,18 @@ def emit_any_requires_grad() -> List[str]:
+    def emit_assert_no_inference_tensor() -> List[str]:
+        tensor_args = []
+        for arg in f.func.arguments.non_out:
+            if isinstance(arg, TensorOptionsArguments):
```
Why is this one skipped separately? The tensor_list check below should be enough no?

ailzhang replied:
Oh it's because `TensorOptionsArguments` doesn't have `argument` attribute. so we need handle it specially anyway.

albanD commented on torch/csrc/autograd/InferenceMode.h:
```diff
@@ -0,0 +1,10 @@
+
+namespace torch { namespace autograd {
+
+using InferenceMode= c10::InferenceMode;
```
nit spacing

albanD commented on torch/csrc/autograd/functions/utils.h:
```diff
@@ -35,6 +35,11 @@ struct ComputeRequiresGrad : IterArgs<ComputeRequiresGrad> {
     }
+    TORCH_CHECK(!tensor.defined() || !tensor.is_view()
+      || static_cast<DifferentiableViewMeta*>(tensor.unsafeGetTensorImpl()->autograd_meta())->get_creation_meta() != CreationMeta::NO_VARIABLE_TYPE_VIEW,
+      "Input contains a view created in InferenceMode without proper grad_fn setup, "
```
I am not sure about these error messages.
Do you have an example of user code that would raise this?

ailzhang replied:
Yup https://github.com/pytorch/pytorch/pull/53343/files#diff-78dfc27f4cd89424e2e342c82969d194f104d138ac70eb1069192735068af21eR260

albanD replied:
I would phrase this slightly differently to be more consistent with this kind of errors in gradmode (https://github.com/pytorch/pytorch/blob/76129c7cdfee6f33861a8b1602326a98bd6c22d0/torch/csrc/autograd/variable.cpp#L510)

This error only happens when GradMode is enabled right?
This error happens for any out of place op? If so, then calling `.clone()` would also throw and error? Do we have a better suggestion?

ailzhang replied:
I replaced it with detach(), I feel this one is a better suggestion?

albanD commented on torch/csrc/autograd/functions/utils.h:
```diff
@@ -54,6 +59,33 @@ inline bool compute_requires_grad(Args&&... args) {
 
+struct AssertNoInferenceTensor : IterArgs<AssertNoInferenceTensor> {
+  using IterArgs<AssertNoInferenceTensor>::operator();
+  void operator()(const at::Tensor& tensor) {
```
This should also handle TensorList no?
Otherwise the checks for things like `cat` won't be correct.

ailzhang replied:
Yea I think so, for example we currently generate code like below which means `compute_requires_grad` handles TensorList. 
```
  Tensor _cat(c10::DispatchKeySet ks, TensorList tensors, int64_t dim) {
    auto tensors_ = unpack(tensors, "tensors", 0);
    auto _any_requires_grad = compute_requires_grad( tensors );
```

albanD replied:
Ho IterArgs handles List and std::vector unpacking by default. Nice!

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -535,6 +536,8 @@ void handle_view_on_rebase(DifferentiableViewMeta* diff_view_meta, bool indirect
                        "version of the function that produced this view or "
                        "don't modify this view inplace.");
       } else {
+        TORCH_INTERNAL_ASSERT(creation_meta != CreationMeta::NO_VARIABLE_TYPE_VIEW,
```
Why is this one internal?

ailzhang replied:
Oh it's because we error out for `NO_VARIABLE_TYPE_VIEW` in the beginning of VariableType kernel. Also tensors with `NO_VARIABLE_TYPE_VIEW` are view of normal tensors which means they have Autograd keys and will definitely go through VariableType kernel first. Thus we shouldn't expect to see NO_VARIABEL_TYPE_VIEW in as_view at all.

albanD replied:
This function can also be called from `.grad_fn` if another view was modified inplace and so a Tensor lazily rebase his history because someone is trying to access its grad_fn

ailzhang replied:
Hmmm right, but access to grad_fn should also have been blocked by the the check in the beginning of VariableType I assume? ;)

albanD reviewed (commented):
The broad lines look quite good.

I skipped the tests for now, will look at these tomorrow.

ezyang commented on c10/core/InferenceMode.h:
```diff
@@ -0,0 +1,67 @@
+   prev_keyset(c10::impl::tls_local_dispatch_key_set()) {
+    // Expected state:
+    //   InferenceMode: no InplaceOrView in included, Autograd in excluded
+    //   NormalMode: InplaceOrView in included, no Autograd in excluded
```
An alternative that feels more symmetric to me is (1) put InplaceOrView on tensors (like Autograd), and (2) add InplaceOrView and Autograd both to excluded when you turn on InferenceMode. I wonder if there is some semantic problem with doing it this way.

ezyang replied:
Is there ever a case where InplaceOrView is excluded but inference mode TLS is not true?

ezyang replied:
> The proposed idea won't work since in this way both your "normal tensor" and "inference tensor" (tensors created in InferenceMode) will look the same (they both have Autograd and InplaceOrView keys).

Hmm, I don't follow. Currently the TensorImpl construction logic is:

```
 if (c10::InferenceMode::is_enabled()) {
    // See Note [Expected TLS state in InferenceMode] for why we don't add Autograd & InplaceOrView keys.
    key_set_ = key_set;
  } else {
    // TODO: Ideally we only add AutogradBackend key when the tensor requires grad.
    //       See Note [Dream: skip VariableType kernel when requires_grad=false]
    DispatchKey k = key_set.highestPriorityBackendTypeId();
    key_set_ = key_set | getAutogradRelatedKeySetFromBackend(k);
  }
```

It seems to me that you would want to *unconditionally* set InplaceOrView, but modulate Autograd on whether or inference mode is enabled. Then a normal tensor will have Autograd and InplaceOrView, but a inference mode tensor will only have InplaceOrView.

ezyang replied:
> Currently no, InplaceOrView is only removed from included when inference mode TLS is true. Are you thinking about we can eliminate the inference mode TLS and check InplaceOrView key instead? ;)

Perhaps. It's a little indirect so it might not be good for understandability, but it certainly would reduce the amount of TLS you have to do.

albanD replied:
> inference mode tensor will only have InplaceOrView.

You don't actually want that. You want Inference Tensor to have that key only if they are a view of a non-inference mode Tensor.
All other inference mode Tensors should not have that key.

ailzhang replied:
Hmmm I'm happy to take that indirection if it reduces a TLS :D.

Commit:
Update on "Implement public API InferenceMode and its error handling"


Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

ezyang commented on c10/core/InferenceMode.h:
```diff
@@ -0,0 +1,65 @@
+
+// A RAII, thread local (!) guard that enables or disables inference mode upon
+// construction, and sets it back to the original value upon destruction.
+struct TORCH_API InferenceMode {
```
If you turn on inference mode, and then you ship some work to a thread in a thread pool, inference mode won't apply inside that thread (because it's TLS). Because you added a new TLS for inference mode, even if dispatch key TLS is propagated, that inference mode TLS won't get propagated unless you add the logic in the relevant spots.

albanD replied:
We have a central TLS state system in c10 that is used to propagate states across threads.
Namely: https://github.com/pytorch/pytorch/blob/db5e5b439c454d657cfa8f08a096cf68e203f2a8/aten/src/ATen/ThreadLocalState.h#L11

You want to add a field there for inference mode and properly update the get and set functions.

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -302,6 +307,8 @@ at::DataPtr PlacementDeleteContext::makeDataPtr(
 
 void TensorImpl::set_requires_grad(bool requires_grad) {
+  // Cannot set requires_grad=true in InferenceMode
+  requires_grad = !c10::InferenceMode::is_enabled() && requires_grad;
```
The choice is between making it an error and ignoring it, it doesn't really feel like there is any in-between. I suppose we *could* put `requires_grad` as an attribute on TensorImpl and then further extend the "no autogradmeta" optimization to apply to leaf nodes with requires_grad=True, which would then make this work (you would error for non-leaf but that's OK).

albanD replied:
Well you could also let this function run as before, just like we do for GradMode.
It allows you to set the requires_grad field to whatever you want, but it is still ignored if you try to do computations based on that Tensor (that would be true for both GradMode and InferenceMode).
I would vote for consistency between the two here.

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,286 @@
+    - Autograd=true, InplaceOrView=true
+    - Autograd=false, InplaceOrView=false
+*/
+void assert_tensor_dispatch_keys(torch::Tensor& x, bool Autograd, bool InplaceOrView) {
```
If these keys are lockstep, wouldn't it be better to just have `bool InferenceMode` in the argument instead?

ezyang commented on c10/core/impl/LocalDispatchKeySet.cpp:
```diff
@@ -101,4 +101,8 @@ void tls_set_dispatch_key_included(DispatchKey x, bool desired_state) {
   }
 }
 
+bool tls_is_dispatch_keyset_excluded(DispatchKeySet ks) {
```
maybe less ambiguous with `is_all_dispatch_keyset_excluded` (the ambiguity is because `any` qualifier is also possible here)

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,286 @@
+
+TEST(InferenceModeTest, TestInferenceTensorCreation) {
+  {
+    InferenceMode guard(true);
```
Idiomatic use is to omit the boolean true, right? Probably better if the tests follow idiomatic use too, people may look to them to see how to use it.

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,286 @@
+    // it'll be silently ignored. This is not an error to make adding InferenceMode
+    // guard to existing code a smooth user experience.
+    torch::Tensor c = torch::ones({1, 2, 3});
+    assert_tensor_dispatch_keys(c, false, false);
```
@ailzhang this gets back to the "abstract model" I mentioned to you earlier. If the intention is that you think of tensors allocated in inference mode to be "inference tensors", and this holds the most explanatory power for when things work versus not work, then we should take this concept seriously and actually refer to it that way (even if only in these tests). So rather than `assert_tensor_dispatch_keys`, you would say `assert_is_inference_tensor` (and maybe seriously consider adding this test into the public API). "Tensors allocated in inference mode are *inference tensors*. Some operations do not work on inference tensors..."

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,286 @@
+  torch::Tensor a = torch::ones({1, 2, 3}).set_requires_grad(true);
+  torch::Tensor b = torch::ones({1, 2, 3}).set_requires_grad(false);
+  torch::Tensor tmp, c;
+  // assignment operator preserves original keyset.
```
This test seems kind of pointless because: "of course assignment operator preserves! It better, because Tensor is a shared pointer..." XD

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,286 @@
+  }
+  // perform backward on `a` should trigger error since `a`'s version has been bumped.
+  ASSERT_THROWS_WITH(out.backward(torch::ones_like(out)),
+    "one of the variables needed for gradient computation has been modified by an inplace operation")
```
This is so beautiful 😭

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,286 @@
+  assert_tensor_dispatch_keys(func_out, false, false);
+
+  torch::Tensor inplace_out = inplace_op(c);  // go through kernels: CPU
+  assert_tensor_dispatch_keys(inplace_out, false, false);
```
This is also a weird thing to test because we expect `inplace_out is c`

albanD commented on c10/core/InferenceMode.h:
```diff
@@ -0,0 +1,65 @@
+    //     torch::Tensor c = a.view({4, 1}); // (*)
+    //     If we don't add InplaceOrView to included set, (*) will skip its as_view
+    //     setup entirely which is wrong.
+    //     By going through InplaceOrView kernel, we can either throw an error
```
Ok, I think you still want to say we one we do here :D

albanD commented on c10/core/InferenceMode.h:
```diff
@@ -0,0 +1,65 @@
+    //     By going through InplaceOrView kernel, we can either throw an error
+    //     or setup as_view accordingly.
+
+    DispatchKeySet included = enabled ? prev_keyset.included_.remove(c10::DispatchKey::InplaceOrView)
```
Yes I was asking about these ones. Sounds good.

albanD commented on torch/script.h:
```diff
@@ -3,6 +3,7 @@
 #include <torch/csrc/api/include/torch/types.h>
 #include <torch/csrc/autograd/generated/variable_factories.h>
 #include <torch/csrc/autograd/grad_mode.h>
+#include <torch/csrc/autograd/InferenceMode.h>
```
cc @glaringlee is that the right place to put this? Should we just put it in `autograd.h` instead?

Commit:
Update on "Implement public API InferenceMode and its error handling"


Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

Commit:
Update on "Implement public API InferenceMode and its error handling"


Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,276 @@
+
+    tmp = torch::ones({1, 2, 3}).set_requires_grad(false);
+    ASSERT_FALSE(tmp.requires_grad());
+    assert_tensor_dispatch_keys(tmp, /*is_inference_tensor=*/true);
```
nit: `ASSERT_TRUE(is_inference_tensor(tmp))`? lol

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -500,7 +500,7 @@ struct TORCH_API ViewInfo {
 /// - DEFAULT is for all other cases
 enum class CreationMeta: uint8_t { DEFAULT, IN_CUSTOM_FUNCTION, MULTI_OUTPUT_NODE,
-                                   NO_GRAD_MODE, MULTI_OUTPUT_SAFE };
+                                   NO_GRAD_MODE, MULTI_OUTPUT_SAFE, NO_VARIABLE_TYPE_VIEW };
```
Don't forget to update the comment above

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,276 @@
+    InferenceMode guard;
+    inference_tensor = torch::ones({1, 2, 3});
+  }
+  // We cannot add c10::autograd_dispatch_keyset to gloablly enabled, because that will
```
nit: globally

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,276 @@
+    inference_tensor = torch::ones({1, 2, 3});
+  }
+  // We cannot add c10::autograd_dispatch_keyset to gloablly enabled, because that will
+  // accidentally call autograd kernel from a backend that might not match tensor input.
```
I don't really understand this sentence. Is it setting up some sort of counterfactual? *Suppose* we added autograd dispatch keyset to globally enabled (word missing here), then... this test would fail? The sentence seems like it is making a statement about how AutogradXLA/AutogradCPU key design means you can't put them in the global set, but that doesn't seem relevant to the test here?

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,276 @@
+    // inplace -> inplace -> view
+    torch::Tensor view_out = view_op(a);
+    assert_tensor_dispatch_keys(view_out, /*is_inference_tensor=*/false);
+    ASSERT_TRUE(view_out.requires_grad());
```
Not only is the behavior equivalent, but if you were to run autograd on this result, it would work, right?

ezyang replied:
No it wouldn't, because the InplaceOrView kernel only would version bump, it wouldn't also update the autograd metadata. So I wonder what actually would happen if you tried to call backward on the mutated tensor later.

ezyang replied:
I'm not sure why you wouldn't mark the resulting view in this situation as `requires_grad=False`. The behavior here is consistent with `no_grad`, but this behavior is controversial https://github.com/pytorch/pytorch/issues/11390 also see #12502. Because inference mode is all about letting implications of things that happen in the inference mode region escape (in the name of performance), it doesn't make sense to me to preserve the old view behavior.

Actually, I might even go a step further, and say that inplacing a `requires_grad=True` tensor inside an inference mode block should set `requires_grad=False`. But I suppose we can cross that bridge later (certainly seems more annoying to implement).

ezyang replied:
I suppose an argument for the current behavior is that InferenceMode is intended to be exactly semantically equivalent to `no_grad`, except that there are more error cases in some situations. However, I checked https://fb.quip.com/JFgeAL0arWCA and didn't see this explicated as an explicit design goal (though I could have missed it!)

ezyang commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,276 @@
+
+TEST(InferenceModeTest, TestMixInferenceAndNormalTensor) {
+  // Note you can only mix inference tensor and normal tensor in functional ops.
+  // FIXME: Inplace & view ops only take one tensor input.
```
Inplace ops can take more than one tensor, even `tensor.add_(tensor)` is a simple example

Commit:
Update on "Implement public API InferenceMode and its error handling"


RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

Commit:
Update on "Implement public API InferenceMode and its error handling"


RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

Commit:
Update on "Implement public API InferenceMode and its error handling"


RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

ailzhang commented:
As discussed with @albanD that a user facing InferenceMode definition will be: InferenceMode is a faster version of NoGradMode, with an additional limitation that you cannot do inplace/view op on tensors created in InferenceMode outside InferenceMode. 

(In this PR mixing inference tensor and normal tensor outside InferenceMode is also forbidden, but we can discuss whether to relax it in later PRs). 

A followup is whether to not allocate VersionCounter or allocate a special VersionCounter for InferenceTensor to allow maximum perf or flexibility. But there isn't use case for it yet so we can address it later.

ezyang commented on c10/core/InferenceMode.cpp:
```diff
@@ -0,0 +1,15 @@
+
+// Invariant:
+//   Normal mode: InplaceOrView in TLS included
+//   InferenceMode: InferenceMode not in TLS included
```
typo s/InferenceMode/InplaceOrView/

ezyang replied:
Also this isn't really an invariant; you're basically defining inference mode to be the state where inplaceorview is not in TLS included

ezyang commented on c10/core/InferenceMode.h:
```diff
@@ -0,0 +1,65 @@
+  static bool is_enabled();
+
+  private:
+    bool prev_mode;
```
This is dead now (no reads from this variable), so you can just delete it.

ezyang commented on c10/core/impl/LocalDispatchKeySet.cpp:
```diff
@@ -6,7 +6,7 @@ namespace c10 {
 namespace impl {
 
 // NB: POD, zero initialized!
```
out of date now

ezyang commented on c10/core/impl/LocalDispatchKeySet.h:
```diff
@@ -121,5 +121,6 @@ C10_API bool tls_is_dispatch_key_excluded(DispatchKey x);
 C10_API void tls_set_dispatch_key_excluded(DispatchKey x, bool desired_state);
 C10_API bool tls_is_dispatch_key_included(DispatchKey x);
 C10_API void tls_set_dispatch_key_included(DispatchKey x, bool desired_state);
+C10_API bool is_all_dispatch_keyset_excluded(DispatchKeySet ks);
```
This queries tls right? Prefix the function with `tls_`

ezyang commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -301,9 +308,8 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
-            creation_meta = ('at::GradMode::is_enabled() ? '
-                             'CreationMeta::DEFAULT : '
-                             'CreationMeta::NO_GRAD_MODE')
+            creation_meta = ('(!at::GradMode::is_enabled() || InferenceMode::is_enabled()) ? '
```
I guess this raises the question again whether or not InferenceMode should also toggle `at::GradMode::is_enabled`. Doing so the easy way will cause an extra TLS change in the InferenceMode guard (although that isn't really a big deal).

ezyang commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -301,9 +308,8 @@ def emit_view_body(fn: NativeFunctionWithDifferentiabilityInfo, var: str) -> Tup
-                             'CreationMeta::DEFAULT : '
-                             'CreationMeta::NO_GRAD_MODE')
+            creation_meta = ('(!at::GradMode::is_enabled() || InferenceMode::is_enabled()) ? '
+                             'CreationMeta::NO_GRAD_FN : CreationMeta::DEFAULT')
```
Is `NO_GRAD_FN` really that accurate a new name for the CreationMeta? To me, this reads like "this as no `grad_fn`", but I don't think the CreationMeta here has any thing to do with whether or not you have a grad fn; for example, if I have a `requires_grad=False` tensor and I take a view on it, IIUC, I'll get a view with `CreationMeta::DEFAULT`, even though the `grad_fn` is empty.

ezyang commented:
So, in this version of the PR + the one stacked on top, the algorithm by which we achieve inference mode is as follows:

* InplaceOrView is in included TLS by default
* InferenceMode removes InplaceOrView from included TLS, adds Autograd to excluded TLS
* During Tensor allocation, if InferenceMode, do not allocate version counter and do not explicitly put Autograd+InplaceOrView key on Tensor. A tensor that is missing the InplaceOrView key is called an "inference tensor"
  * Important side fact: when constructing view tensor, key set is exactly replicated from source tensor. So even if we don't explicitly put an Autograd key on a tensor, the Autograd key may be indirectly propagated via key set
* In this PR: If inference tensor is found in VariableType kernel, raise an error
  * After next PR: It is not allowed to bump a missing VC. You can query current version but it will always report -1.
* During view creation, if InferenceMode, set CreationMeta to NO_GRAD_FN
* If InplaceOrView kernel is called while autograd is NOT in the exclude set (indicating that the Autograd kernel was bypassed), throw an error

The correctness conditions for the algorithm are:

1. (Correctness) If a tensor is saved for backwards (i.e., VC read), no alias of the tensor may be mutated in a way that bypasses version counter bump
2. (Performance) If in inference mode, must not setup view metadata when doing views on inference tensors (aka tensors guaranteed not to alias with tensors saved for backwards)
3. (Performance) If in inference mode, must not do VC bumps for inference tensors (aka as above)

There are probably some other correctness conditions. (more)

ezyang commented:
Some high level comments after writing down the algorithm:

* One of the reasons why I've been advocating for merging this and the second PR is because this PR does a very imperfect job at ensuring inference tensors don't get saved for backwards. VariableType is codegenned to handle it, but custom autograd functions are not handled at all (users can still save tensors for backwards there). In general, someone could implement an entirely homebrew AD system based on VCs being accurate, and you'd have no way of knowing about it. That's why a simpler and more correct way to avoid correctness problems is to just error if you try to read the VC on an inference tensor. (That being said, it seems like there's a problem with the second PR because you are erroring on bump rather than read, so we need to talk about that, but let's shelve this point for now).
* I think all three of the correctness conditions are satisfied, although for (1) it can be a little hard to see. First the easy ones: (2) and (3) are true because inference tensors don't have InplaceOrView, so they bypass inplace / view setup when InplaceOrView isn't also in the ambient TLS. For (1), we have to establish the invariant that inference tensors (tensors without InplaceOrView and Autograd) are guaranteed not to alias with normal tensors. There are two ways this could be violated: an inference tensor view constructed from a normal tensor, and a normal tensor view constructed from an inference tensor. In order:
  * An inference tensor view can only be constructed in inference mode, as outside of inference mode we force the addition of Autograd+InplaceOrView. By test case, we see that view on normal tensor results in a normal tensor. The reason this happens is because of key set propagation; although Autograd+InplaceOrView is not explicitly set, we still inherit the key set from the parent
  * Taking a view on an inference tensor results in an error, because of the "is autograd in exclude set" test. The InplaceOrView kernel is guaranteed to still be called because it is in the local TLS. This is also verified in test.
* For follow up: it seems like now that InplaceOrView is in the default TLS set, `no_grad` can be implemented simply as "exclude Autograd in TLS", as all the unconditional logic is now moved out

ezyang reviewed (approved):
Approving because I think the algorithm works. @ailzhang I mentioned to you that I want a little more juice out of the RFC (and the in code comments, I suppose); I'll do some editing on the RFC to try to give you an idea of what I'm going for, but you don't have to block this PR on it.

Commit:
Update on "Implement public API InferenceMode and its error handling"


RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

Commit:
Update on "Implement public API InferenceMode and its error handling"


RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

Commit:
Update on "Implement public API InferenceMode and its error handling"


RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

ezyang commented:
Was reviewing the spec some more and I think we can probably do some more safe relaxations:

* Taking a view on an inference mode tensor outside of inference mode is safe, you just need to produce an inference tensor in this case as well (currently this is an error)
* Inplace modifying an inference mode tensor outside of inference mode is safe and you can skip the VC bump

Commit:
Update on "Implement public API InferenceMode and its error handling"


RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

Commit:
Update on "Implement public API InferenceMode and its error handling"


RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

Commit:
Update on "Implement public API InferenceMode and its error handling"


RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

bhosmer commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,391 @@
+    ASSERT_TRUE(is_inference_tensor(c));
+
+    torch::Tensor tmp = torch::ones({1, 2, 3}).set_requires_grad(true);
+    ASSERT_TRUE(tmp.requires_grad()); // requires_grad is silently ignored when it's an inference tensor.
```
Did we decide there was a reason we wanted to silently ignore, like code reuse? Seems clear it would be saner devex to actually error if you set requires_grad true on an inference tensor.

ezyang replied:
The code got edited above, but just to be super explicit: it's not silently ignored; we DO set `requires_grad=True` (just like in no grad mode); the idea to not error is so that code that initializes parameters can run in inference mode without triggering errors

ailzhang replied:
Yea code reuse is the main reason. And another reason is inference tensor can actually have requires_grad=true (and gradients) as shown in the second PR. :D

bhosmer reviewed (approved):
This LGTM per Ed's analysis here and offline conversation 😁  one note about setting `requires_grad`, but if we want to do anything about that it can definitely wait for a followup.

Commit:
Update on "Implement public API InferenceMode and its error handling"


RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

Commit:
Update on "Implement public API InferenceMode and its error handling"


RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

albanD commented on c10/core/InferenceMode.h:
```diff
@@ -0,0 +1,63 @@
+    //   NormalMode: InplaceOrView in included, Autograd not in excluded
+    //
+    // 1. When InferenceMode is enabled, Autograd dispatch keys are excluded
+    //    but not InplaceOrView key.
```
nit: This mostly repeats the comment above but in a more confusing way.
Should we remove the comment above and add the exact state here?

albanD commented on c10/core/InferenceMode.h:
```diff
@@ -0,0 +1,63 @@
+struct TORCH_API InferenceMode {
+  InferenceMode(bool enabled=true): prev_keyset(c10::impl::tls_local_dispatch_key_set()) {
+    // Note [Expected TLS state in InferenceMode]:
+    //   InferenceMode: InplaceOrView not in included, Autograd in excluded
```
nit: you can mention here that InplaceOrView is never in the excluded set and Autograd is never in the included set.

albanD commented on c10/core/InferenceMode.h:
```diff
@@ -0,0 +1,63 @@
+    //       a = in.view({1, 4});
+    //     }
+    //     torch::Tensor c = a.view({4, 1}); // (*)
+    //     If we don't add InplaceOrView to included set, (*) will skip its as_view
```
Is this still true now that we have `c` that is also an inference Tensor?
Or is that change in another PR?

ailzhang replied:
Yea that's for the second PR. Will change it there!

ailzhang replied:
Yea that's for the second PR. Will change it there!

albanD commented on c10/core/InferenceMode.cpp:
```diff
@@ -0,0 +1,10 @@
+namespace c10 {
+
+bool InferenceMode::is_enabled() {
+  return !c10::impl::tls_is_dispatch_key_included(DispatchKey::InplaceOrView);
```
If we're planning on calling this function from perf critical code (which I think we do since we call this in every TensorImpl creation), I would suggest we actually save the boolean argument given to the InferenceMode constructor and just return that here. That will avoid the TLS read.

ailzhang replied:
Oh actually I think we cannot avoid the TLS read here. since this function is static that we call with `InferenceMode::is_enabled()`, it's not associated to any InferenceMode object. If we saved some boolean during constructor, it'll have to be a TLS boolean as well which makes no difference in perf.

albanD replied:
Ho it is static, my bad.
TLS read it is then :)

albanD commented on c10/core/TensorImpl.h:
```diff
@@ -563,6 +563,11 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
 
+  // Inference tensor doesn't have autograd or InplaceOrView key.
+  bool is_inference_tensor() {
+    return !key_set_.has(c10::DispatchKey::InplaceOrView) && (key_set_ & c10::autograd_dispatch_keyset).empty();
```
Since we read both keys here, should we have a TORCH_INTERNAL_CHECK that makes sure that we don't have any invalid state where one is set but not the other?

albanD commented on tools/autograd/gen_inplace_or_view_type.py:
```diff
@@ -335,6 +342,19 @@ def emit_inplace_or_view_body(fn: NativeFunctionWithDifferentiabilityInfo) -> Li
     else:
         api_name = sig_group.signature.name()
+    # Throws if a InplaceOrView kernel is hit without Autograd being disabled.
+    # There're two valid cases calling InplaceOrView:
```
I think this comment is nice but it is a bit weird to have a comment that explains what this code is *not* doing :p

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -658,6 +662,15 @@ def emit_any_requires_grad() -> List[str]:
 
+    def emit_assert_no_inference_tensor() -> List[str]:
+        tensor_arg_names = []
+        for arg in f.func.arguments.tensor_args:
```
Does that include TensorList?

Given the code below it looks like no but it should.

Can we add a test where we use `cat()` and make sure this code is properly triggered?

ailzhang replied:
Yup tensor args already include TensorList and added a `cat()` test case :D

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -509,7 +510,10 @@ void handle_view_on_rebase(DifferentiableViewMeta* diff_view_meta, bool indirect
+    } else if (creation_meta == CreationMeta::INFERENCE_MODE) {
+      msg = c10::str("A view was created in inference mode and ", modified_obj, " modified inplace in normal mode.");
     } else {
+      TORCH_INTERNAL_ASSERT_DEBUG_ONLY(creation_meta == CreationMeta::NO_GRAD_MODE);
```
Actually I am not sure this will be accurate if you do (the old code had the same issue and was not reporting the right error message AFAIU):
```
a = torch.rand(10)
b = a.view_as(a)
b.requires_grad_()
```

The creation meta here will be default but you don't have a grad_fn.

ailzhang replied:
I removed this line but the example actually doesn't hit here, it's blocked by https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/variable.cpp#L413
and we won't hit this line if creation_meta=Default.

albanD commented on test/cpp/api/grad_mode.cpp:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  void assert_tensor_creation_meta(torch::Tensor& x, torch::autograd::CreationMeta creation_meta) {
+    ASSERT_EQ(static_cast<torch::autograd::DifferentiableViewMeta*>(x.unsafeGetTensorImpl()->autograd_meta())->get_creation_meta(), creation_meta);
```
nit: you might wat to add a check that the autograd meta is not null and has a backward view before this no? Or we are ok with this segfaulting if we pass in a Tensor that does not verify these?

albanD commented on test/cpp/api/grad_mode.cpp:
```diff
@@ -0,0 +1,86 @@
+TEST(GradModeTest, TestRequiresGradViewOpExiting) {
+  for (bool requires_grad: {true, false}) {
+    torch::Tensor s = torch::ones({1, 2, 3}).set_requires_grad(requires_grad);
+    torch::Tensor a = s + 2;
```
nit you can clone in these cases to avoid the magic number

albanD commented on test/cpp/api/grad_mode.cpp:
```diff
@@ -0,0 +1,86 @@
+
+    {
+      torch::AutoGradMode mode(false);
+      view_out = view_op(a);  // go through kernels: InplaceOrView, CPU
```
Not sure about the comment? This goes through VariableType as well

albanD commented on test/cpp/api/grad_mode.cpp:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    tmp = view_op(view_out);
+    ASSERT_EQ(tmp.requires_grad(), requires_grad);
```
nit you can add another `assert_tensor_creation_meta(tmp, torch::autograd::CreationMeta::NO_GRAD_MODE);` here? To ensure the chaining happened properly?

albanD commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,391 @@
+  which listed below, even though these tensors are created in InferenceMode, they're
+  normal tensors:
+    - tensors which are view of normal tensors.
+  Note output of inplace ops on normal tensors(which is the input itself).
```
Not sure what this means. The last 3 lines here are a bit confusing (maybe remove the list if there is a single element in that list).

albanD commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,391 @@
+  bool has_Autograd = ks.has(c10::DispatchKey::AutogradCPU);
+  bool has_InplaceOrView = ks.has(c10::DispatchKey::InplaceOrView);
+  // They must be either both true or false.
+  bool is_inference_tensor = !has_Autograd && !has_InplaceOrView && x.is_leaf();
```
Why do you check if it is a leaf here??

albanD commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,391 @@
+}
+
+void assert_tensor_creation_meta(torch::Tensor& x, torch::autograd::CreationMeta creation_meta) {
+  ASSERT_EQ(static_cast<torch::autograd::DifferentiableViewMeta*>(x.unsafeGetTensorImpl()->autograd_meta())->get_creation_meta(), creation_meta);
```
Same comment as the no grad version about possible segfault

albanD commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,391 @@
+
+void assert_TLS_states(bool inference_mode) {
+  ASSERT_EQ(InferenceMode::is_enabled(), inference_mode);
+  ASSERT_EQ(c10::impl::tls_is_dispatch_keyset_excluded(c10::autograd_dispatch_keyset), inference_mode);
```
nit you can actually check here that autograd is never included and inplace or view is never excluded?

albanD commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,391 @@
+    ASSERT_TRUE(is_inference_tensor(c));
+
+    torch::Tensor tmp = torch::ones({1, 2, 3}).set_requires_grad(true);
+    ASSERT_TRUE(tmp.requires_grad()); // requires_grad=true on inference tensor is no-op.
```
not sure what that comment means.
setting that field is not a noop as you actually properly set it.
But it will be ignored when performing computations.

albanD commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,391 @@
+      ASSERT_EQ(a.requires_grad(), requires_grad);
+
+      // inplace -> inplace
+      inplace_op(a);  // go through kernels: InplaceOrView, CPU
```
Why do you do this twice?

ailzhang replied:
no particular reason, just thinking that we might want to have a test on chaining ops. ;)

albanD commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,391 @@
+      // inplace -> inplace -> view
+      torch::Tensor view_out = view_op(a);
+      ASSERT_FALSE(is_inference_tensor(view_out));
+      // WHY??
```
Because the view's requires_grad is its own or his base's.

albanD commented on test/cpp/api/inference_mode.cpp:
````diff
@@ -0,0 +1,391 @@
+      //     auto impl = c10::make_intrusive<TensorImpl>(
+      //     Storage(self.storage()), self.key_set(), self.dtype());
+      //   ```
+      //   In addition, these view output tensors although are normal in the sense
````
missing a subject in that sentence?

albanD commented on test/cpp/api/inference_mode.cpp:
````diff
@@ -0,0 +1,391 @@
+      //   ```
+      //   In addition, these view output tensors although are normal in the sense
+      //   that it has both Autograd and InplaceOrView keys, they're still special
+      //   since they'll have CreationMeta::NO_GRAD_MODE. In other words they behave
````
INFERENCE_MODE here no?

albanD commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,391 @@
+      assert_tensor_creation_meta(tmp, CreationMeta::INFERENCE_MODE);
+      ASSERT_FALSE(is_inference_tensor(tmp));
+      ASSERT_TRUE(tmp.requires_grad());
+      ASSERT_TRUE(tmp.is_leaf());
```
Shall we check here that a's version is the same as tmp's version?

albanD commented on test/cpp/api/inference_mode.cpp:
```diff
@@ -0,0 +1,391 @@
+    }
+
+    tmp = view_op(view_out);
+    ASSERT_FALSE(is_inference_tensor(view_out));
```
You should check the creation meta here to make sure it is still INFERENCE_MODE

albanD reviewed (commented):
One case that I don't see mentioned here is:

```
a = torch.rand(10, requires_grad=True)
with inference_mode():
    b = a.view_as(a)
with torch.no_grad():
    c = b.view_as(b)
c += 1
```
I guess this is fine, but just want to make sure https://github.com/pytorch/pytorch/blob/dfc7fa03e5d33f909b9d7853dd001086f5d782a0/torch/csrc/autograd/variable.h#L508-L515 handles that case properly.


Also after looking at all the examples, I am now convinced that we should present inference mode as no grad but more restrictive. As most of the behavior is actually the same.

Commit:
Update on "Implement public API InferenceMode and its error handling"


RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

Commit:
Update on "Implement public API InferenceMode and its error handling"


RFC: https://github.com/pytorch/rfcs/pull/17

Differential Revision: [D26973911](https://our.internmc.facebook.com/intern/diff/D26973911)

[ghstack-poisoned]

albanD reviewed (approved):
LGTM!
Thanks for all the updates!
